src/data/complex/R_or_C.lean

```diff
@@ -0,0 +1,603 @@
+(mul_im_ax : ∀ z w : K, im (z * w) = re z * im w + im z * re w)
+(conj_re_ax : ∀ z : K, re (conj z) = re z)
+(conj_im_ax : ∀ z : K, im (conj z) = -(im z))
+(conj_I_ax : conj I = -I)
+(norm_sq_eq_def_ax : ∀ (z : K), ∥z∥^2 = (re z) * (re z) + (im z) * (im z))
+(mul_im_I_ax : ∀ (z : K), (im z) * im I = im z)
+(inv_def_ax : ∀ (z : K), z⁻¹ = conj z * of_real ((∥z∥^2)⁻¹))
+(div_I_ax : ∀ (z : K), z / I = -(z * I))
+
+namespace is_R_or_C
+
+variables {K : Type*} [normed_field K] [algebra ℝ K] [is_R_or_C K]
+local notation `𝓚` := @is_R_or_C.of_real K _ _ _
+
+lemma of_real_alg : ∀ x : ℝ, 𝓚 x = x • (1 : K) :=
+  λ x, by rw [←mul_one (𝓚 x), smul_coe_mul_ax]
+
+@[simp] lemma re_add_im (z : K) : 𝓚 (re z) + 𝓚 (im z) * I = z := is_R_or_C.re_add_im_ax z
+@[simp] lemma of_real_re : ∀ r : ℝ, re (𝓚 r) = r := is_R_or_C.of_real_re_ax
+@[simp] lemma of_real_im : ∀ r : ℝ, im (𝓚 r) = 0 := is_R_or_C.of_real_im_ax
+@[simp] lemma mul_re : ∀ z w : K, re (z * w) = re z * re w - im z * im w :=
+  is_R_or_C.mul_re_ax
+@[simp] lemma mul_im : ∀ z w : K, im (z * w) = re z * im w + im z * re w :=
+  is_R_or_C.mul_im_ax
+
+theorem ext_iff : ∀ {z w : K}, z = w ↔ re z = re w ∧ im z = im w :=
+λ z w, { mp := by { rintro rfl, cc },
+         mpr := by { rintro ⟨h₁,h₂⟩, rw [←re_add_im z, ←re_add_im w, h₁, h₂] } }
+
+theorem ext : ∀ {z w : K}, re z = re w → im z = im w → z = w :=
+by { simp_rw ext_iff, cc }
+
+
+lemma zero_re : re (𝓚 0) = (0 : ℝ) := by simp only [of_real_re]
+@[simp] lemma zero_im : im (𝓚 0) = 0 := by rw [of_real_im]
+lemma of_real_zero : 𝓚 0 = 0 := by rw [of_real_alg, zero_smul]
+
+@[simp] lemma of_real_one : 𝓚 1 = 1 := by rw [of_real_alg, one_smul]
+@[simp] lemma one_re : re (1 : K) = 1 := by rw [←of_real_one, of_real_re]
+@[simp] lemma one_im : im (1 : K) = 0 := by rw [←of_real_one, of_real_im]
+
+@[simp] theorem of_real_inj {z w : ℝ} : 𝓚 z = 𝓚 w ↔ z = w :=
+{ mp := λ h, by { convert congr_arg re h; simp only [of_real_re] },
+  mpr := λ h, by rw h }
+
+
+@[simp] lemma bit0_re (z : K) : re (bit0 z) = bit0 (re z) := by simp [bit0]
+@[simp] lemma bit1_re (z : K) : re (bit1 z) = bit1 (re z) :=
+  by simp only [bit1, add_monoid_hom.map_add, bit0_re, add_right_inj, one_re]
+@[simp] lemma bit0_im (z : K) : im (bit0 z) = bit0 (im z) := by simp [bit0]
+@[simp] lemma bit1_im (z : K) : im (bit1 z) = bit0 (im z) :=
+  by simp only [bit1, add_right_eq_self, add_monoid_hom.map_add, bit0_im, one_im]
+
+@[simp] theorem of_real_eq_zero {z : ℝ} : 𝓚 z = 0 ↔ z = 0 :=
+  by rw [←of_real_zero]; exact of_real_inj
+
+@[simp] lemma of_real_add (r s : ℝ) : 𝓚 (r + s) = 𝓚 r + 𝓚 s :=
+  by apply (@is_R_or_C.ext_iff K _ _ _ (𝓚 (r + s)) (𝓚 r + 𝓚 s)).mpr; simp
+
+@[simp] lemma of_real_bit0 (r : ℝ) : 𝓚 (bit0 r : ℝ) = bit0 (𝓚 r) :=
+ext_iff.2 $ by simp [bit0]
+
+@[simp] lemma of_real_bit1 (r : ℝ) : 𝓚 (bit1 r : ℝ) = bit1 (𝓚 r) :=
+ext_iff.2 $ by simp [bit1]
+
+/-- Note: This can be proven by norm_num once K is proven to be of characteristic zero below. -/
```

Comment:
This probably shouldn't be a doc string since it's not too useful on hover.
```suggestion
/- Note: This can be proven by `norm_num` once K is proven to be of characteristic zero below. -/
```